eslint/eslint-patch/src/eslint-bulk-suppressions/bulk-suppressions-patch.ts

```diff
@@ -15,265 +14,277 @@ interface ISuppression {
+    return node.key.name;
+  } else if (Guards.isNormalClassPropertyDefinitionWithAnonymousExpressionAssigned(node)) {
+    return node.key.name;
+  } else if (Guards.isNormalAssignmentPatternWithAnonymousExpressionAssigned(node)) {
+    return node.left.name;
+  } else if (Guards.isNormalMethodDefinition(node)) {
+    return node.key.name;
+  } else if (Guards.isTSEnumDeclaration(node)) {
+    return node.id.name;
+  } else if (Guards.isTSInterfaceDeclaration(node)) {
+    return node.id.name;
+  } else if (Guards.isTSTypeAliasDeclaration(node)) {
+    return node.id.name;
+  }
 }
 
 type NodeWithParent = TSESTree.Node & { parent?: TSESTree.Node };
 
 function calculateScopeId(node: NodeWithParent | undefined): string {
   const scopeIds: string[] = [];
   for (let current: NodeWithParent | undefined = node; current; current = current.parent) {
-    const scopeIdForASTNode: string | null = getNodeName(current);
-    if (scopeIdForASTNode !== null) scopeIds.unshift(scopeIdForASTNode);
+    const scopeIdForASTNode: string | undefined = getNodeName(current);
+    if (scopeIdForASTNode !== undefined) {
+      scopeIds.unshift(scopeIdForASTNode);
+    }
   }
 
-  if (scopeIds.length === 0) return '.';
-
-  return '.' + scopeIds.join('.');
+  if (scopeIds.length === 0) {
+    return '.';
+  } else {
+    return '.' + scopeIds.join('.');
+  }
 }
 
-/**
- * @throws Throws an error if the command to retrieve the root path fails.
- * @returns The root path of the monorepo.
- */
-export function getGitRootPath(): string {
-  const result: SpawnSyncReturns<string> = spawnSync('git', ['rev-parse', '--show-toplevel'], {
-    encoding: 'utf-8'
-  });
-  if (result.status !== 0) throw new Error(`get root path failed`);
-  return result.stdout.toString().trim();
-}
+const eslintrcPathByFileOrFolderPath: Map<string, string> = new Map();
 
-export const GitRootPath: string = getGitRootPath();
+function findEslintrcFolderPathForNormalizedFileAbsolutePath(normalizedFilePath: string): string {
+  const cachedFolderPathForFilePath: string | undefined =
+    eslintrcPathByFileOrFolderPath.get(normalizedFilePath);
+  if (cachedFolderPathForFilePath) {
+    return cachedFolderPathForFilePath;
+  }
+  const normalizedFileFolderPath: string = normalizedFilePath.substring(
+    0,
+    normalizedFilePath.lastIndexOf('/')
+  );
 
-function findEslintrcDirectory(fileAbsolutePath: string): string {
+  const pathsToCache: string[] = [normalizedFilePath];
+  let eslintrcFolderPath: string | undefined;
   for (
-    let currentDir: string = fileAbsolutePath;
-    currentDir.startsWith(GitRootPath);
-    currentDir = path.dirname(currentDir)
-  )
-    if (['.eslintrc.js', '.eslintrc.cjs'].some((eslintrc) => fs.existsSync(path.join(currentDir, eslintrc))))
-      return currentDir;
-  throw new Error('Cannot locate eslintrc');
-}
-
-function validateSuppressionsJson(json: IBulkSuppressionsJson): json is IBulkSuppressionsJson {
```

Comment:
It's uncomfortable to be reading files without ANY validation.  For example, imagine a support ticket from someone who's confused because ESLint keeps crashing with this error:
```
Uncaught TypeError: Cannot read properties of undefined (reading 'length')
```
...because line 12,345 of **.eslint-bulk-suppressions.json** says `scopeID` instead of `scopeId`, a bad Git merge resolution, etc.

Does this validation really have a nontrivial performance cost if we're only reading the file once?  If so, maybe we could validate during `eslint-bulk suppress` and `eslint-bulk prune` but not actual linting.

Reply:
It'd suffice to validate the file when it is loaded; as long as it only loads once per process that should be fine.